### PR TITLE
Fix cherry-pick

### PR DESCRIPTION
**Pull request by bpeabody**
This changes addresses many outstanding cherry-pick problems, including:

* `--continue` and `--abort`: https://github.com/twosigma/git-meta/issues/462
* picking meta-commits that introduce multiple submodule commits in one
  repo: https://github.com/twosigma/git-meta/issues/361
* failures in libgit2 touching deleted submodules:
  https://github.com/twosigma/git-meta/issues/360
* handling additions:
  https://github.com/twosigma/git-meta/issues/337
* state cleanup https://github.com/twosigma/git-meta/issues/224

Key improvements:

* Track that a cherry-pick is in-progress to support --continue and
  --abort
* Stop using `libgit2` cherry-pick in the meta-repo.  This lets us avoid
  doing the expensive "cacheSubmodules" call in the meta-repo and gives
  us more control over what happens.  We also avoid any potential
  submodule bugs re. cherry-picking in libgit2 this way.  Using libgit2
  cherry-pick didn't buy us a whole lot in the meta-repo anyway.
* Use 3-way rebase in the submodules instead of cherry-pick so that we
  can get all changes introduced by commit.
* Minor stability improvements to merge and rebase.
* More consistent checks that we are in a good state to initiate
  operations, e.g., that there are no existing conflicts or cherry-picks
  before starting a merge.
* Better handling of conflicts in general
* Better testing of rebase.  Factored out some parts to use with
  cherry-pick and wrote separate tests for them.
* Performance improvements to operations that were reading the shas of
  all submodules when they actually needed only a subset.

Additionally, this change lays the groundwork to go back and make
further improvements to `merge` and `rebase` (e.g., to make `-i`) should
we choose to.

Derived implementation changes:

* Introduced a type: `CherryPick` for recording a cherry-pick
* Support for reading and writing cherry-pick state in
  `cherry_pick_fileUtil.js`.
* Test framework support for cherry-picks
* Test framework support for conflicts

### CODE REVIEW

File: node/lib/cmd/cherry_pick.js [hunk 1]
```diff
@@ -73,29 +88,56 @@ exports.configureParser = function (parser) {
...
         }
         else {
-            console.log(`Cherry-picking commit ${colors.green(result.id())}.`);
-            let commit = yield repo.getCommit(result.id());
-            yield CherryPick.cherryPick(repo, commit);
+            const id = annotated.id();
+            console.log(`Cherry-picking commit ${colors.green(id.tostrS())}.`);
+            const  commit = yield repo.getCommit(id);
+            const result = yield CherryPickUtil.cherryPick(repo, commit);
+            if (null !== result.errorMessage) {
```
Comment (novalis) [inline node/lib/cmd/cherry_pick.js]: Do we need to store this and the remaining  commits in .git/sequencer/todo?

File: node/lib/util/cherry_pick.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    equal(rhs) {
+        assert.instanceOf(rhs, CherryPick);
+        return this.d_originalHead === rhs.d_originalHead &&
+            this.d_picked === rhs.d_picked;
+    }
+}
+
+CherryPick.prototype.toString = function () {
+    return `Conflict(originalHead=${this.d_originalHead}, \
```
Comment (novalis) [inline node/lib/util/cherry_pick.js]: s/Conflict/CherryPick/

File: node/lib/util/cherry_pick_file_util.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+const CherryPick = require("./cherry_pick");
+
+/**
+ * This module contains methods for accessing files that pertain to in-progress
+ * cherry-picks.
+ */
+
+const metaCherryPickDir = "META_CHERRY_PICK";
+const originalHeadFile = "ORIG_HEAD";
+const pickFile = "PICK";
```
Comment (novalis) [inline node/lib/util/cherry_pick_file_util.js]: why this instead of .git/sequencer?

File: node/lib/util/cherry_pick_file_util.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    const root = path.join(gitDir, metaCherryPickDir);
+    yield fs.mkdir(root);
+    yield fs.writeFile(path.join(root, originalHeadFile),
+                       cherryPick.originalHead + "\n");
+    yield fs.writeFile(path.join(root, pickFile),
+                       cherryPick.picked + "\n");
+});
+
+/**
+ * Remove files related to a meta merge.
```
Comment (novalis) [inline node/lib/util/cherry_pick_file_util.js]: s/merge/cherry-pick/

Comment (novalis) [reply node/lib/util/cherry_pick_file_util.js]: That error implies that this function is copypasta -- would it be possible to extract a common method between these?

File: node/lib/util/cherry_pick_util.js [hunk 5]
```diff
@@ -0,0 +1,566 @@
+/*
+ * Copyright (c) 2016, Two Sigma Open Source
```
Comment (novalis) [inline node/lib/util/cherry_pick_util.js]: 2018

File: node/lib/util/cherry_pick_util.js [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+const SubmoduleConfigUtil = require("./submodule_config_util");
+const SubmoduleUtil       = require("./submodule_util");
+const TreeUtil            = require("./tree_util");
+const UserError           = require("./user_error");
+
+/**
+ * Change the specified `submodules` in the specified index.  If a name maps to
+ * a `Submodule`, update it in the specified `index` in the specified `repo`
+ * and if that submodule is open, reset its HEAD, index, and worktree to
+ * reflect that commit.  Otherwise, if it maps to `null`, remove it and deinit
```
Comment (novalis) [inline node/lib/util/cherry_pick_util.js]: If it's open and has local changes, document what this does.

File: node/test/util/read_repo_ast_util.js [hunk 7]
```diff
@@ -1490,6 +1498,54 @@ describe("readRAST", function () {
...
+        const r = start.repo;
+
+        // Switch to master
+
+        yield r.checkoutBranch("master");
+
+        const head = yield r.getHeadCommit();
+        const sha = head.id().tostrS();
+
+        const cherryPick = new CherryPick(sha, sha);
```
Comment (novalis) [inline node/test/util/read_repo_ast_util.js]: do you mean head, sha instead of sha, sha?

File: node/test/util/repo_status.js [hunk 8]
```diff
@@ -974,6 +1053,7 @@ describe("RepoStatus", function () {
             workdir: { x: FILESTATUS.ADDED },
             rebase: new Rebase("a", "4", "b"),
             merge: new Merge("a", "2", "2"),
+            cherryPick: new CherryPick("a", "2", "2"),
```
Comment (novalis) [inline node/test/util/repo_status.js]: extra arg

File: node/lib/util/cherry_pick_file_util.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+});
+
+/**
+ * Write the specified `cherryPick` to the specified `gitDir`.  The behavior is
+ * undefined if there is already a cherry-pick recorded in `gitDir`.
+ *
+ * @param {String}     gitDir
+ * @param {CherryPick} cherryPick
+ */
+exports.writeCherryPick = co.wrap(function *(gitDir, cherryPick) {
```
Comment (novalis) [inline node/lib/util/cherry_pick_file_util.js]: It looks like this is never called outside of tests -- is that what you intend?

File: node/lib/cmd/cherry_pick.js [hunk 10]
```diff
@@ -73,29 +88,56 @@ exports.configureParser = function (parser) {
...
+        }
+        yield CherryPickUtil.abort(repo);
+        return;                                                       // RETURN
+    }
+
+    if (0 === args.commits.length) {
+        throw new UserError("No commits to cherry-pick.");
+    }
+
     for (let i = 0; i < args.commits.length; ++i) {
```
[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/cmd/cherry_pick.js]: I wish JavaScript worked that way.

File: node/lib/cmd/cherry_pick.js [hunk 11]
```diff
@@ -73,29 +88,56 @@ exports.configureParser = function (parser) {
...
+            throw new UserError(`\
+Could not resolve ${colors.red(commitish)} to a commit.`);
         }
         else {
-            console.log(`Cherry-picking commit ${colors.green(result.id())}.`);
-            let commit = yield repo.getCommit(result.id());
-            yield CherryPick.cherryPick(repo, commit);
+            const id = annotated.id();
+            console.log(`Cherry-picking commit ${colors.green(id.tostrS())}.`);
+            const  commit = yield repo.getCommit(id);
```
[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/cmd/cherry_pick.js]: got it

File: node/lib/cmd/cherry_pick.js [hunk 12]
```diff
@@ -73,29 +88,56 @@ exports.configureParser = function (parser) {
...
+    }
+
+    if (args.abort) {
+        if (0 !== args.commits.length) {
+            throw new UserError("Cannot specify commits with '--abort'.");
+        }
+        yield CherryPickUtil.abort(repo);
+        return;                                                       // RETURN
+    }
+
```
[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/cmd/cherry_pick.js]: It's checked in `CherryPickUtil.abort` (and tested for)

File: node/lib/cmd/cherry_pick.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (novalis) [reply node/lib/cmd/cherry_pick.js]: Wait, isn't that what "for/of" is supposed to do? I know for/in doesn't work.

File: node/lib/cmd/cherry_pick.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (novalis) [reply node/lib/cmd/cherry_pick.js]: No, in the non-abort, non-continue case, before starting a new cherry-pick.

File: node/lib/cmd/cherry_pick.js (hunk 1 above)
Comment (bpeabody) [reply node/lib/cmd/cherry_pick.js]: State management is done in `CherryPickFileUtil`. I'm not using the sequencer yet. There's no support in libgit2 for it that I can see, and I'd actually prefer that norma git commands not see my state (so we don't run into problems like what we have today with people running `git rebase --abort` on git-meta rebases.

However, I would like to replicate the sequencer technique and support, e.g., `rebase -I` at some point -- but I had to draw a line on this change at some point.

File: node/lib/util/cherry_pick_file_util.js (hunk 3 above)
Comment (bpeabody) [reply node/lib/util/cherry_pick_file_util.js]: I would like to do that in a future iteration. Maybe we can discuss tomorrow.

File: node/lib/util/cherry_pick_file_util.js (hunk 4 above)
Comment (bpeabody) [reply node/lib/util/cherry_pick_file_util.js]: Yes, but I don't think it would be a useful refactoring; both are just calling `rimraf`.

File: node/test/util/read_repo_ast_util.js (hunk 7 above)
Comment (bpeabody) [reply node/test/util/read_repo_ast_util.js]: No, it's writing out a cherry-pick that's from and to itself. More exhaustive testing of reading and writing the cherry-pick info is done in `CherryPickFileUtil`; this is just checking that the read APIs are in place.

File: node/lib/util/cherry_pick_file_util.js (hunk 9 above)
Comment (bpeabody) [reply node/lib/util/cherry_pick_file_util.js]: It's called on line 451 of `util/cherry_pick_util.js` when a cherry-pick is started.

File: node/lib/cmd/cherry_pick.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/cmd/cherry_pick.js]: Oh, I was not aware of `for of`!

File: node/lib/cmd/cherry_pick.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/cmd/cherry_pick.js]: Oh yes, in that case `ensureReady` catches it.

**Comment by bpeabody** [comment]
Fixed most of the problems so far.

File: node/lib/util/cherry_pick_util.js [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+    assert.instanceOf(commit, NodeGit.Commit);
+
+    const status = yield StatusUtil.getRepoStatus(metaRepo);
+    StatusUtil.ensureReady(status);
+
+    // First, perform sanity checks to see if the repo is in a state that we
+    // can pick in and if `commit` is something that we can pick.
+
+    if (!status.isDeepClean(false)) {
+        // TODO: Git will refuse to run if there are staged changes, but will
```
[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/util/cherry_pick_util.js]: That's definitely doable. It becomes a little tricky when it comes to abort/continue as it means we'll need to track which submodules are being picked... even then, continue is hard. What if the user had to make changes to a submodule that's not being picked to resolve a problem (like a build breakage) caused by the pick?

File: node/lib/util/status_util.js [hunk 14]
```diff
@@ -314,6 +346,63 @@ exports.getSubmoduleStatus = co.wrap(function *(repo,
...
+                our: null,
+                their: null,
+            };
+            conflicted[path] = obj;
+        }
+        return obj;
+    }
+    const entries = index.entries();
+    const STAGE = RepoStatus.STAGE;
+    for (let i = 0; i < entries.length; ++i) {
```
Comment (novalis) [inline node/lib/util/status_util.js]: for/of?

File: node/lib/util/submodule_util.js [hunk 15]
```diff
@@ -346,6 +351,9 @@ exports.getSubmoduleChangesFromDiff = function (diff) {
                     result[path] = new SubmoduleChange(
                                                  delta.oldFile().id().tostrS(),
                                                  newFile.id().tostrS());
+                } else if (!allowMetaChanges && path !== modulesFileName) {
+                    throw new UserError(`\
+Modification to meta-repo file ${colors.red(path)} is not supported.`);
```
[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/util/submodule_util.js]: No, it works fine with additions and removals (at least as far as my unit tests indicate).

[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/util/submodule_util.js]: What it won't do is a cherry-pick of a commit that manipulates a file in the meta-repo that is not a submodule and not `.gitmodules`.

File: node/lib/util/submodule_config_util.js [hunk 16]
```diff
@@ -558,3 +558,19 @@ exports.writeConfigText = function (urls) {
     }
     return result;
 };
+
+/**
+ * Write, to the configuration file, the specified `urls` in the specified
```
[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/util/submodule_config_util.js]: yes, I'll change it to say that

File: node/lib/util/status_util.js [hunk 17]
```diff
@@ -314,6 +346,63 @@ exports.getSubmoduleStatus = co.wrap(function *(repo,
...
+        }
+    }
+    const result = {};
+    const COMMIT = NodeGit.TreeEntry.FILEMODE.COMMIT;
+    for (let name in conflicted) {
+        const c = conflicted[name];
+
+        // Ignore the conflict if it's just between submodule SHAs
+
+        if (null === c.our ||
```
[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/util/status_util.js]: |expression  |  null | COMMIT | BLOB|
|------------|------|----------|------|
|null === c.our | true | false | false|
|null === c.our OR COMIT !== c.our | true | false | true|

**Comment by bpeabody** [comment]
Made another round of update.

File: node/lib/util/status_util.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (novalis) [reply node/lib/util/status_util.js]: |expression  |  null | COMMIT | BLOB|
|------------|------|----------|------|
|null === c.our | true | false | false|
|null === c.our OR COMMIT !== c.our | true | false | true|
|COMMIT !== c.our | true | false | true|

File: node/lib/util/cherry_pick_util.js [hunk 18]
```diff
@@ -0,0 +1,569 @@
...
+ */
+exports.abort = co.wrap(function *(repo) {
+    assert.instanceOf(repo, NodeGit.Repository);
+
+    const cherry = yield CherryPickFileUtil.readCherryPick(repo.path());
+    if (null === cherry) {
+        throw new UserError("No cherry-pick in progress.");
+    }
+    const commit = yield repo.getCommit(cherry.originalHead);
+    yield Reset.reset(repo, commit, Reset.TYPE.HARD);
```
Comment (novalis) [inline node/lib/util/cherry_pick_util.js]: I think we can start a cherry-pick in an unclean state, right?  That is, this reset might blow away local changes Also, let's say you start a cherry-pick, then get distracted and start making changes in some other submodule.  It seems like we shouldn't delete those changes?

File: node/lib/util/status_util.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (bpeabody) [reply node/lib/util/status_util.js]: Yeah, I'm a moron. But at least I got to use markdown tables.

File: node/lib/util/cherry_pick_util.js (hunk 18 above)
Comment (bpeabody) [reply node/lib/util/cherry_pick_util.js]: No, we will not let you start a pick in an unclean state. Partially because of this issue, and partially because it's extremely difficult to do so (we'd need some sort of dry-run that libgit2 rebase doesn't provide). I have a TODO about allowing it to run in an unclean state on line 421 of `util/cherry_pick_util.js`.

Now, re. files that may have been modified during the cherry-pick for whatever reason, I have a TODO on line 68 of `cmd/cherry_pick.js` where I tell the users that the command blows away all their changes.

**Comment by bpeabody** [comment]
I think I've resolved everything except for the issue you brought up with the HARD reset. @novalis, let's discuss this in-person tomorrow.

File: node/lib/util/cherry_pick_util.js (hunk 18 above)
Comment (novalis) [reply node/lib/util/cherry_pick_util.js]: I somehow missed the code that prevents starting in an unclean state -- can you point me to it?

Comment (bpeabody) [reply node/lib/util/cherry_pick_util.js]: from line 420 of `util/cherry_pick_util.js`

```JavaScript
    if (!status.isDeepClean(false)) {
        // TODO: Git will refuse to run if there are staged changes, but will
        // attempt a cherry-pick if there are just workdir changes.  We should
        // support this in the future, but it basically requires us to dry-run
        // the rebases in all the submodules, and I'm uncertain how to do that
        // at the moment.

        throw new UserError(`\
The repository has uncommitted changes.  Please stash or commit them before
running cherry-pick.`);
    }
```

Comment (novalis) [reply node/lib/util/cherry_pick_util.js]: Oh, I even commented on it. OK, good.

**Comment by bpeabody** [comment]
I just pushed changes to fix a couple of problems I found doing manual testing:
1. Missed an edge case in `computeChanges` where addition of a submodule on both sides was incorrectly diagnosed.
2. When a pick failed due to a conflict I was not flushing the index. Due to timing reasons, my unit test didn't notice.
